common/logger/src/struct_log.rs

```diff
@@ -94,13 +94,18 @@ impl Write for TcpWriter {
         // Attempt to write, and if it fails clear underlying stream
         // This doesn't guarantee a message cut off mid send will work, but it does guarantee that
         // we will connect first
-        let mut stream = self.stream.as_ref().unwrap();
-        let result = stream.write(buf);
-        if result.is_err() {
-            self.stream = None;
+        if let Some(mut stream) = self.stream.as_ref() {
+            let result = stream.write(buf);
+            if result.is_err() {
+                self.stream = None;
+            }
+            result
+        } else {
+            Err(io::Error::new(
+                io::ErrorKind::NotConnected,
+                "Can't write to stream, not connected",
+            ))
```

Comment:
can we replace this with the functor example I gave in the previous PR?

https://github.com/libra/libra/pull/6410#discussion_r501445867

Reply:
I'm wondering why functional over imperative? The problem with rust is that you can write both, and that it ends up being a personal preference. I'd be more interested in a principled approach where we advise one style over the other.

Reply:
Well look how much more verbose this is versus what was provided, that's one principle I use. Another principle is to avoid nesting. Functional has no statement based nesting whereas this has if within an if.

Reply:
I'm trying to do this:
  self.stream
        .as_mut()
        .ok_or_else(|| io::Error::new(io::ErrorKind::NotConnected,"Can't write to stream, not connected"))
        .and_then(|stream|  stream.write(buf).map_err(|e| { self.stream = None; e }));

but I'm getting this

mismatched types
  expected enum `std::result::Result<usize, std::io::Error>`
found unit type `()`rustc(E0308)
struct_log.rs(88, 8): implicitly returns `()` as its body has no tail or `return` expression

That's because the expected return value is a result<usize> but we're not returning in that functional statement. I tried looking online on how to do it but nothing worked. do you have any suggestions?

Reply:
I've also tried:
```
        self.stream
        .as_mut()
        .ok_or_else(|| io::Error::new(io::ErrorKind::NotConnected,"Can't write to stream, not connected"))
        .and_then(|stream|  { let result = stream.write(buf).map_err(|e| { self.stream = None; e }) ; result } );
```
but it also has the same problem

Reply:
> Well look how much more verbose this is versus what was provided

I think the trade-off is verbosity vs readability. So I don't find it that clear cut.

Reply:
@xvschneider I don't think you can write `self.X` in the closure you pas to `map_err`. You also don't return anything to your function (due to the `;` at the end of your statement)

Reply:
this worked no problem:
```
        self.stream
            .as_mut()
            .ok_or_else(|| io::Error::new(io::ErrorKind::NotConnected, "No stream"))
            .and_then(|stream| stream.write(buf))
            .map_err(|e| {
                self.stream = None;
                e
            })
```

I think your issue was adding the ";" at the end of the call.

let's get a 3rd vote, cc: @bmwill

Reply:
Yeah I'm ok with this, I believe Greg already signed off on this in the last iteration. Lets just make sure to either squash our commits or apply the `bors-squash` label

Reply:
I can't add the label, if you have permissions can you add bors-squash?